src/util/AbstractMemory/Memory.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (July of 2023,
+// schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <cmath>
+#include <concepts>
+#include <cstddef>
+#include <string>
+#include <type_traits>
+
+#include "util/AbstractMemory/CalculationUtil.h"
+#include "util/Exception.h"
+
+namespace ad_utility {
+
+/*
+An abstract class, that represents an amount of memory.
+Note: Literals for easier usage of the class were defined after under the class
+definition.
+*/
+class Memory {
+  // Because of `sizeof` we know, that any size of memory, in bytes, can be
+  // encoded as a `size_t` in cpp.
+  size_t memoryInBytes_;
+
+ public:
+  // Very simple custom constructor.
+  Memory(const size_t& amountOfMemoryInBytes = 0)
+      : memoryInBytes_{amountOfMemoryInBytes} {}
+
+  // Default assignment operator.
+  Memory& operator=(const Memory&) = default;
+  Memory& operator=(Memory&&) = default;
+
+  // Custom assignment operator, so that we can assign user literals.
+  Memory& operator=(const size_t& amountOfMemoryInBytes);
+  Memory& operator=(size_t&& amountOfMemoryInBytes);
+
+  /*
+  Return the internal memory amount in the wanted memory unit format.
+  For example: If the internal memory amount is 1024 bytes, than `kilobytes()`
+  would return `1.0`.
+  */
+  size_t bytes() const;
+  double kilobytes() const;
+  double megabytes() const;
+  double gigabytes() const;
+  double terabytes() const;
+  double petabytes() const;
+
+  /*
+  Return the string representation of the internal memory amount in the
+  biggest memory unit, that is equal to, or smaller than, the internal memory
+  amount.
+  Example: 1024 bytes would be returned as `"1 KB"`.
+  */
+  std::string asString() const;
```

Comment:
If you have `asString` then also provide `operator<<` for ostream, we will often want it.

Reply:
Done, see [6d47cf7](https://github.com/ad-freiburg/qlever/pull/1032/commits/6d47cf76eb184d839ef4b9a80b002b0d9c6d3360).